File: backend-nest/src/common/middleware/payload-size.middleware.ts
import {
  Injectable,
  NestMiddleware,
  PayloadTooLargeException,
} from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { Request, Response, NextFunction } from 'express';

/**
 * Middleware to validate payload size limits for bulk operations.
 * Prevents memory exhaustion attacks by limiting request body size.
 */
@Injectable()
export class PayloadSizeMiddleware implements NestMiddleware {
  constructor(private readonly configService: ConfigService) {}

  use(req: Request, res: Response, next: NextFunction) {
    // Default to 1MB limit for bulk operations
    const maxSize = this.configService.get<number>(
      'BULK_OPERATIONS_MAX_PAYLOAD_SIZE',
      1048576,
    );

    // Check if this is a bulk operations endpoint
    const isBulkOperation = req.path.includes('/bulk-operations');

    if (isBulkOperation && req.headers['content-length']) {
      const contentLength = parseInt(req.headers['content-length']);

      if (contentLength > maxSize) {
        throw new PayloadTooLargeException(
          `Payload size (${contentLength} bytes) exceeds maximum allowed size (${maxSize} bytes) for bulk operations`,
        );
      }
    }

    next();
  }
}

File: backend-dotnet/src/Pulpe.Api/Api/Middleware/PayloadSizeMiddleware.cs
namespace Pulpe.Api.Api.Middleware;

public class PayloadSizeMiddleware
{
    private readonly RequestDelegate _next;
    private const long MaxBulkPayloadBytes = 1024 * 1024; // 1MB

    public PayloadSizeMiddleware(RequestDelegate next)
    {
        _next = next;
    }

    public async Task InvokeAsync(HttpContext context)
    {
        if (context.Request.Path.Value?.Contains("/bulk-operations") == true)
        {
            var contentLength = context.Request.ContentLength;
            if (contentLength.HasValue && contentLength.Value > MaxBulkPayloadBytes)
            {
                context.Response.StatusCode = StatusCodes.Status413RequestEntityTooLarge;
                return;
            }
        }

        await _next(context);
    }
}
